lib/Mojo/Log.pm

```diff
@@ -0,0 +1,363 @@
+
+=head2 message
+
+  $log->on(message => sub ($log, $level, @lines) {...});
+
+Emitted when a new message gets logged.
+
+  $log->on(message => sub ($log, $level, @lines) { say "$level: ", @lines });
+
+=head1 ATTRIBUTES
+
+L<Mojo::Log> implements the following attributes.
+
+=head2 color
+
+  my $bool = $log->color;
+  $log     = $log->color($bool);
+
+Colorize log messages with the levels C<warn>, C<error> and C<fatal> using L<Term::ANSIColor>, defaults to the value of
+the C<MOJO_LOG_COLOR> environment variables. Note that this attribute is B<EXPERIMENTAL> and might change without
+warning!
+
+=head2 format
+
+  my $cb = $log->format;
+  $log   = $log->format(sub {...});
+
+A callback for formatting log messages.
+
+  $log->format(sub ($time, $level, @lines) { "[2018-11-08 14:20:13.77168] [28320] [info] I ♥ Mojolicious\n" });
+
+=head2 handle
+
+  my $handle = $log->handle;
+  $log       = $log->handle(IO::Handle->new);
+
+Log filehandle used by default L</"message"> event, defaults to opening L</"path"> or C<STDERR>.
+
+=head2 history
+
+  my $history = $log->history;
+  $log        = $log->history([[time, 'debug', 'That went wrong']]);
+
+The last few logged messages.
+
+=head2 level
+
+  my $level = $log->level;
+  $log      = $log->level('debug');
+
+Active log level, defaults to C<trace>. Available log levels are C<trace>, C<debug>, C<info>, C<warn>, C<error> and
+C<fatal>, in that order.
+
+=head2 max_history_size
+
+  my $size = $log->max_history_size;
+  $log     = $log->max_history_size(5);
+
+Maximum number of logged messages to store in L</"history">, defaults to C<10>.
+
+=head2 path
+
+  my $path = $log->path
+  $log     = $log->path('/var/log/mojo.log');
+
+Log file path used by L</"handle">.
+
+=head2 short
+
+  my $bool = $log->short;
+  $log     = $log->short($bool);
+
+Generate short log messages without a timestamp but with journald log level prefix, suitable for systemd environments,
+defaults to the value of the C<MOJO_LOG_SHORT> environment variables.
+
+=head1 METHODS
+
+L<Mojo::Log> inherits all methods from L<Mojo::EventEmitter> and implements the following new ones.
+
+=head2 append
+
+  $log->append("[2018-11-08 14:20:13.77168] [28320] [info] I ♥ Mojolicious\n");
+
+Append message to L</"handle">.
+
+=head2 capture
+
+  my $messages = $log->capture;
+  my $messages = $log->capture('debug');
+
+Capture log messages for as long as the returned object exists, useful for testing log messages.
+
+  # Test your log messages
+  my $messages = $log->capture('trace');
+  $log->fatal('Something very bad happened');
+  $log->trace('Just some debug information');
+  like $messages, qr/Something very bad happened/, 'logs contain fatal message';
+  like $messages->[-1], qr/Just some debug information/, 'trace message was last';
+  undef $messages;
+
+=head2 context
+
+  my $new = $log->context('[extra]', '[information]');
+
+Construct a new child L<Mojo::Log> object that will include context information with every log message.
```

Comment:
Would it make sense to allow context to propagate to the next context call?  $c->log->info reporting [request_id] is nice, and then you might want to add additional context to that...  What's the best way to do that?  And can it be made more easy?